app/views/play/cinematic/ByteCode/AbstractCommand.js

```diff
@@ -0,0 +1,82 @@
+import * as Promise from 'bluebird'
+
+Promise.config({
+  cancellation: true
+})
+
+/**
+ * These symbols can be exported in order to call the protected AbstractCommand methods.
+ */
+export const run = Symbol('private run method')
+export const cancel = Symbol('cancellation symbol')
+
+/**
+ * AbstractCommand is the abstract base class for objects that will be run in the
+ * BytecodeThunkRunner.
+ *
+ * @abstract
+ */
+export default class AbstractCommand {
+  constructor () {
+    if (this.constructor === AbstractCommand) {
+      throw new Error('This is an abstract class that you must extend.')
+    }
+  }
+
+  /**
+   * This method should not be overriden.
+   * It checks that the promise returned from run is valid, and also remembers
+   * the promise so it can be cancelled.
+   */
+  [run] () {
+    this.promise = this.run()
+    if (!this.promise) {
```

Comment:
This check could be better. We could check that `.then` method exists.

We could also sanitize all returned values from `run` by passing them through `Promise.resolve()`.
This would better hide the command implementation. I like the `run` method having a strict promise return interface.

This makes it very clear as to what promise is passed into the `cancel` method that you can override.